### PR TITLE
feat: support dag and steps level scheduling constraints. Fixes: #12568

### LINKED ISSUE #12568: Workflow ignoring nodeSelector and tolerations when defined in WorkflowTemplate
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I can confirm the issue exists when I tested with `:latest`
- [X] I have searched existing issues and could not find a match for this bug
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/main/docs/CONTRIBUTING.md))

### What happened/what did you expect to happen?

I defined a WorkflowTemplate containing `nodeSelector` and `tolerations`. When I manually submit the WorkflowTemplate from the UI both fields are ignored and the workflow is running on any available node in the cluster.

### Version

v3.5.4

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
apiVersion: argoproj.io/v1alpha1
kind: WorkflowTemplate
metadata:
  name: benchmarks
spec:
  entrypoint: main
  serviceAccountName: workflow
  templates:
  - dag:
      tasks:
      - arguments:
          parameters:
          - name: msg
            value: 'hello'
        name: benchmark
        template: benchmark
    name: main
    nodeSelector:
      pool: workflows
    tolerations:
    - key: pool
      operator: Equal
      value: workflows
  - inputs:
      parameters:
      - name: msg
    name: benchmark
    script:
      command:
      - python
      image: python:latest
      source: |
        print("{{inputs.parameters.msg}}")
```

### Logs from the workflow controller

```text
kubectl logs -n argo deploy/workflow-controller | grep ${workflow}
```

### Logs from in your workflow's wait container

```text
kubectl logs -n argo -c wait -l workflows.argoproj.io/workflow=${workflow},workflow.argoproj.io/phase!=Succeeded
```

### PR DESCRIPTION
**Pull request by shuangkun**
Fixes #12568

### Motivation

I want to Take effect dag and steps level scheduling constraints. Not just workflow level and template level.

### Modifications

Find the pod's bounaryID's Template and add the scheduling constraints to pod if exists.

### Verification

ut and e2e test

### CODE REVIEW

File: workflow/controller/workflowpod.go [hunk 1]
```diff
@@ -815,6 +826,28 @@ func addSchedulingConstraints(pod *apiv1.Pod, wfSpec *wfv1.WorkflowSpec, tmpl *w
 	}
 }
 
+// GetBoundaryTemplate get a template through the node's BoundaryID.
+func (woc *wfOperationCtx) GetBoundaryTemplate(nodeName string) (*wfv1.Template, error) {
```
Comment (Joibel) [inline workflow/controller/workflowpod.go:830]: Can something like this function also be used from https://github.com/argoproj/argo-workflows/blob/e00abd1ded9e8f4787093b84f6cb26ee0870ec7c/workflow/controller/operator.go#L2740

The two bits of code are very similar to each other, and it's good to extract common functionality whilst we're doing this kind of work.

This function could take a *NodeStatus instead

File: workflow/controller/workflowpod.go [hunk 2]
```diff
@@ -764,22 +764,33 @@ func (woc *wfOperationCtx) addMetadata(pod *apiv1.Pod, tmpl *wfv1.Template) {
 }
 
 // addSchedulingConstraints applies any node selectors or affinity rules to the pod, either set in the workflow or the template
-func addSchedulingConstraints(pod *apiv1.Pod, wfSpec *wfv1.WorkflowSpec, tmpl *wfv1.Template) {
+func (woc *wfOperationCtx) addSchedulingConstraints(pod *apiv1.Pod, wfSpec *wfv1.WorkflowSpec, tmpl *wfv1.Template, nodeName string) {
```
Comment (Joibel) [inline workflow/controller/workflowpod.go:767]: Potentially switching to nodeID instead of nodeName makes some of this slightly simpler

**Review by Joibel — CHANGES_REQUESTED** [review]
This is good stuff.

I'd just like to rework some of the function calls here to allow a bit more code reuse.

File: workflow/controller/workflowpod.go (hunk 2 above)
Comment (shuangkun) [reply workflow/controller/workflowpod.go:767]: OK, thanks I will.

File: workflow/controller/workflowpod.go (hunk 1 above)
Comment (shuangkun) [reply workflow/controller/workflowpod.go:830]: OK， I will try it.

Comment (shuangkun) [reply workflow/controller/workflowpod.go:830]: I have made some reuses, please take a look，Thanks.

File: workflow/controller/workflowpod.go (hunk 2 above)
Comment (shuangkun) [reply workflow/controller/workflowpod.go:767]: I find it better to use nodeName to find the Boundary because we have GetNodeByName.

Comment (Joibel) [reply workflow/controller/workflowpod.go:767]: Ok, what you've done with two functions below makes sense. I like it.

**Review by Joibel — APPROVED** [review]
LGTM, thanks @shuangkun

**Review by juliev0 — APPROVED** [review]

**Pull request merged**